lib/SILOptimizer/Transforms/SILMem2Reg.cpp

```diff
@@ -2124,6 +2144,47 @@ void MemoryToRegisters::removeSingleBlockAllocation(AllocStackInst *asi) {
   }
 }
 
+void MemoryToRegisters::collectStoredValues(AllocStackInst *asi,
+                                            StackList<SILValue> &owned,
+                                            StackList<SILValue> &guaranteed) {
+  for (auto *use : asi->getUses()) {
+    auto *user = use->getUser();
+    if (auto *si = dyn_cast<StoreInst>(user)) {
+      owned.push_back(si->getSrc());
+    } else if (auto *sbi = dyn_cast<StoreBorrowInst>(user)) {
+      guaranteed.push_back(sbi->getSrc());
+    }
+  }
+}
+
+void MemoryToRegisters::canonicalizeValueLifetimes(
+    StackList<SILValue> &owned, StackList<SILValue> &guaranteed) {
```

Comment:
What is the difference between calling canonicalize here v/s running a round of copy propagation ? I thought we canonicalize a value in a pass, only if it enables additional optimization, like in silcombine ?

Reply:
Yeah, adding a run of CopyPropagation was the first solution that came to mind after noticing that there were values whose copy-extended lifetime could newly be canonicalized after Mem2Reg but weren't before OME runs.  This alternative came out of discussion with @atrick.  

It might be reasonable to have as a principle that a pass which introduces a new value ought to be responsible for canonicalizing it's lifetime.  That could be used to justify canonicalizing the lifetimes of the new phis.

For the stored values, we'd have to say something like "passes that see that they've created opportunities to canonicalize copy-extended lifetimes can do so".  It seems kind of a shame to toss in a pass to the pipeline to rediscover opportunities for canonicalization that were known about already.

If we think about the CanonicalizeOSSALifetime and CanonicalizeBorrowScope as utilities that to be called from anywhere rather than as a part of the CopyPropagation pass, this perspective might become more appealing.

Reply:
Yeah, you are  right it feels wasteful to throw away newly discovered opportunities. But this also seems to make passes complex and might make bisection slightly harder. A flag to disable the canonicalization might address bisection however.

Reply:
Good call!

Reply:
> Yeah, you are right it feels wasteful to throw away newly discovered opportunities. But this also seems to make passes complex and might make bisection slightly harder.

These are all important points. In this case, we've already established elsewhere that we need a very simple and efficient OSSA canonicalization utility that runs on specific values after a pass that exposes copies.

We may have discussed before that there are two kinds of borrow canonicalization: (1) a trivial utility that runs on SILFunction argument and basically does that same thing as owned canonicalization and (2) a fairly involved analysis and rewriting of borrow scopes. We should make it more clear in the API that only the first form should be called as a cleanup utility and that the second form should only run during the dedicated CopyPropagation pass.